update: add info for the new returnsResultOf contract (#5520)
JetBrains/kotlin-web-site: docs/topics/unused-return-value-checker.md

}
```

> > You can also use the [experimental `returnsResultOf()` contract](whatsnew24.md#improve-unused-result-checks-for-higher-order-functions) to enable the checker to distinguish between meaningful and ignorable unused results from higher-order functions that return the result of a lambda.

Reviewer: This note looks very out of context. Why didn't you add a whole paragraph, like in what's new?

```
## Return value checker and higher order functions

To better convey semantic of higher-order functions, Kotlin has`returnsResultOf()` contract to improve the [unused return value checker](unused-return-value-checker.md).

This contract enables the checker to distinguish between ignorable and meaningful unused results from higher-order functions that return the result of a lambda, such as the `.let()` extension function.

> Kotlin contracts are [Experimental](components-stability.md#stability-levels-explained). To opt > in, add the `@OptIn(ExperimentalContracts::class)` annotation when declaring a function with > a contract.
>
{style=”warning”}

To use this feature, add `returnsResultOf()` to the function's `contract` block:

```kotlin
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.contract

@OptIn(ExperimentalContracts::class)
inline fun <T, R> T.myLet(block: (T) -> R): R {
    contract {
        returnsResultOf(block)
    }
    return block(this)
}
```

Here's an example that uses a custom `myLet()` function with a nullable value:

```kotlin
fun handleNullablePackageName(s: String?, sb: StringBuilder) {
    // The checker doesn't report a warning because append() has an ignorable return value
    s?.myLet { sb.append(it) }

    // The checker reports a warning because the returned String is unused
    s?.myLet { "kotlin.$it" }
    // Unused return value.
}
```

```

Author: as discussed on Slack let's add it as a separate section then but cover the compiler option as well.

}
```

## Check for unused results in higher-order functions

> Some higher-order functions, such as the `.let()` extension function, return the result of a lambda.

Reviewer 2: ```suggestion
Some higher-order functions, such as the `let` scope function, return the result of a lambda.
```
You mean the scope function, right?

> To check for unused results for these functions, add the [Experimental](components-stability.md#stability-levels-explained) `returnsResultOf()` contract to the function's `contract` block.

Reviewer 2: ```suggestion
To check for unused lambda results of higher-order functions, add the [Experimental](components-stability.md#stability-levels-explained) `returnsResultOf()` contract to the function's contract.
```
In the past, we've just referred to it as "contract" so I think we can continue doing the same. Also, what do you think about this for a beginning instead?

Author: I like the start (bit less connected to the previous sentence, but feels more natural) — For the contract part, I have no strong opinion, tbh I find the original a bit clearer, and it avoids repeating the word contract: add contract to the function's contract.

Reviewer 2: I mainly want to avoid saying block because we currently use that for parts of the Gradle build file 🤔

> Kotlin contracts are Experimental. To opt in, add the `@OptIn(ExperimentalContracts::class)` annotation when declaring a function with a contract.
>
{style="warning"}

Here's an example:

```kotlin
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.contract

@OptIn(ExperimentalContracts::class)

> inline fun <T, R> T.myLet(block: (T) -> R): R {

Reviewer 2: Can we use a better name for the `.myLet()` function? Like `transform()`?

Author: Hmm 🤔 I think something like `transform()` would be misleading, and this way it contrasts with the previous example, the let() scope function. we could perhaps call it `customLet()`, but I think `.myLet()` works here.

Reviewer 2: `customLet()` would be fine :) `myLet()` sounds more like a placeholder than a meaningful name that gives some idea what the purpose of the function is.

    contract {
        returnsResultOf(block)
    }
    return block(this)
}
```

You can then use a function with this contract, such as `.myLet()`, to check if the lambda result is used:

```kotlin

> fun handleNullablePackageName(s: String?, sb: StringBuilder) {

Reviewer 2: Can you use better names for the arguments? Like maybe `packageName` and `builder`?

Author: sure 👍 good idea 🙏

>     // The checker doesn't report a warning because append() has an ignorable return value

Reviewer 2: ```suggestion
    // The checker doesn't report a warning because the return value of the append() function can be ignored
```
Ignorable isn't a common word so how about this instead?

    s?.myLet { sb.append(it) }

>     // The checker reports a warning because the returned String is unused

Reviewer 2: ```suggestion
    // The checker reports a warning because the returned string is unused
```

    s?.myLet { "kotlin.$it" }
}
```

> The `returnsResultOf()` contract requires a separate compiler option to opt in, and using it produces pre-release binaries that Kotlin compiler versions earlier than 2.4.0 can't read.

Reviewer 2: ```suggestion
The `returnsResultOf()` contract requires a separate compiler option to opt in. Be aware that using it produces pre-release binaries that Kotlin compiler versions earlier than 2.4.0 can't read.
```
I think this is important enough to become a warning.

> To enable it in your project, add the following compiler option to your build file:

Reviewer 2: ```suggestion
To opt in for your project, add the following compiler option to your build file:
```

}
```

## Check for unused results in higher-order functions

Some higher-order functions, such as the `.let()` extension function, return the result of a lambda.

> To check for unused results in these functions, add the [Experimental](components-stability.md#stability-levels-explained) `returnsResultOf()` contract to the function's `contract` block.

Reviewer: I think "in" is slightly wrong: in `s?.myLet { "kotlin.$it" }` we report `myLet` is unused, so we check "outside" the lambda function. Maybe just say "for these functions"? I don't know better preposition.

Author: 🤔 That is a great point, thank you. for these functions works perfectly imho 💯

> Kotlin contracts are Experimental. To opt in, add the `@OptIn(ExperimentalContracts::class)` annotation when declaring a function with a contract.
>
{style="warning"}

Here's an example:

```kotlin
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.contract

@OptIn(ExperimentalContracts::class)
inline fun <T, R> T.myLet(block: (T) -> R): R {
    contract {
        returnsResultOf(block)
    }
    return block(this)
}
```

> You can then use a function with this contract, such as `.myLet()`, to check the lambda result for unused results:

Reviewer: ```suggestion
You can then use a function with this contract, such as `.myLet()`, to check if the lambda result is properly used:
```

result...results sounds too repetetive

Author: sure that version works well as well 👍 — I'd suggest removing the word *properly* unless we explain what properly means here, it adds noise

```kotlin
fun handleNullablePackageName(s: String?, sb: StringBuilder) {
    // The checker doesn't report a warning because append() has an ignorable return value
    s?.myLet { sb.append(it) }

    // The checker reports a warning because the returned String is unused
    s?.myLet { "kotlin.$it" }

>     // Unused return value.

Reviewer: Obsolete comment?

Author: We can remove one of them, although one of them (above the line) states what happens, and the other (below the line) shows the output. Or am I missing something? Was something changed here? 🤔

Reviewer: I didn't have a clue what it meant until you explained it to me. Do we usually do this with warnings? I think I've never seen this approach before (including this very same document)

Author: I put it there for clarity, but it's not necessary. If you think it creates confusion, let's remove the output part. 👍

Reviewer: The fact that this is the only place throughout the document where you did that indeed creates confusion

}
```

## Check for unused results in higher-order functions

> Some higher-order functions, such as the `let()` scope function, return the result of a lambda.

Reviewer 2: ```suggestion
Some higher-order functions, such as the `let` scope function, return the result of a lambda.
```
We don't use parentheses when talking about scope functions :)
